File: 159__Longest_Substring_with_At_Most_Two_Distinct_Characters.py
class Solution(object):
    def lengthOfLongestSubstringTwoDistinct(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) <= 2:
            return len(s)
        result = 2
        n = len(s)
        i, j = 0, 1
        dic = {s[0] : 1}
        while j < n:
            if s[j] not in dic:
                if len(dic) < 2:
                    dic[s[j]] = 1
                    j += 1
                    result = max(result, j - i)
                else:
                    while i < j:
                        dic[s[i]] -= 1
                        i += 1
                        if dic[s[i - 1]] == 0:
                            del dic[s[i - 1]]
                            break
            else:
                dic[s[j]] += 1
                j += 1
                result = max(result, j - i)

        return result

File: Java/159__Longest_Substring_with_At_Most_Two_Distinct_Characters.java
class Solution {
    public int lengthOfLongestSubstringTwoDistinct(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int l = 0, r = 0, result = 0;
        while (r < s.length()) {
            map.put(s.charAt(r), r++);
            if (map.size() > 2) {
                int index = Collections.min(map.values());
                map.remove(s.charAt(index));
                l = index + 1;
            }
            result = Math.max(result, r - l);
        }
        return result;
    }
}
